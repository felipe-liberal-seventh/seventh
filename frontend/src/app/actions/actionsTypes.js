export const GET_MONITOR = 'GET_MONITOR';
export const REFRESH_HOSTS = 'REFRESH_HOSTS';
export const GET_HOSTS = 'GET_HOSTS';
export const DELETE_HOST = 'DELETE_HOST';
export const ADD_HOST = 'ADD_HOST';


