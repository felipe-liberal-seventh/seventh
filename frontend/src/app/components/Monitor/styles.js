import styled from 'styled-components';

export const containerChart = styled.div`
    margin: 0 20px;
    padding: 0 20px;
    
    @media(max-width: 600px) {
        justify-content: center;
    }
    `
