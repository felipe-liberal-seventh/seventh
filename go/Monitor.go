package main

import (
	"fmt"
	"log"
	"time"
)

//Monitor struct que encapsula o result das metrics
type Monitor struct {
	Metric   []*Metric `json:"results,omitempty"`
	Metadata *Metadata `json:"metadata,omitempty"`
}

//Metric struct que traz os resultados da consulta ao host
type Metric struct {
	ID           int       `json:"id,omitempty"`
	hostID       int       `json:"hostID,omitempty"`
	monitorDate  time.Time `json:"monitorDate,omitempty"`
	url          string    `json:"url,omitempty"`
	status       string    `json:"statusCod,omitempty"`
	statusCod    int       `json:"status,omitempty"`
	timeResponse int64     `json:"timeResponse,omitempty"`
}

//Metadata detalhes da consulta
type Metadata struct {
	TotalCount int32 `json:"total_count,omitempty"`
}

//NewMonitor cria um novo host
func NewMonitor(metric Metric) int64 {
	db := connect()
	defer db.Close()

	//Init transaction
	tx, _ := db.Begin()

	// Insert some data into table.
	sqlStatement, e := db.Prepare("INSERT INTO monitors(hostID, monitorDate, url, statusCod, status, timeResponse) VALUES (?,?,?, ?, ?, ?);")
	if e != nil {
		log.Fatal(e)
		return 0
	}
	res, err := sqlStatement.Exec(&metric.hostID, &metric.monitorDate, &metric.url, &metric.statusCod, &metric.status, &metric.timeResponse)

	fmt.Println("ms:", metric.timeResponse)

	if err != nil {
		tx.Rollback()
		log.Fatal(err)
		return 0
	}
	id, _ := res.LastInsertId()
	tx.Commit()

	return id
}

// CountMonitor retorna o total de registro
// func CountMonitor() (count int) {
// 	db := connect()
// 	defer db.Close()
// 	rows, _ := db.Query(`SELECT COUNT(*) AS count FROM monitors`)
// 	for rows.Next() {
// 		err := rows.Scan(&count)
// 		if err != nil {
// 			log.Fatal(err)
// 		}
// 	}
// 	return count
// }
