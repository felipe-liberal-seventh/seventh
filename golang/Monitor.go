package main

import (
	"fmt"
	"log"
	"time"
)

//Monitor struct que encapsula o result das metrics
type Monitor struct {
	Metric   []*Metric `json:"results,omitempty"`
	Metadata *Metadata `json:"metadata,omitempty"`
}

//Metric struct que traz os resultados da consulta ao host
type Metric struct {
	ID           int64         `json:"id,omitempty"`
	hostId       int64         `json:"hostId,omitempty"`
	monitorDate  time.Time     `json:"monitorDate,omitempty"`
	url          string        `json:"url,omitempty"`
	status       string        `json:"statusCod,omitempty"`
	statusCod    int           `json:"status,omitempty"`
	timeResponse time.Duration `json:"timeResponse,omitempty"`
	created_at   time.Time     `json:"created_at,omitempty"`
	updated_at   time.Time     `json:"updated_at,omitempty"`
}

//Metadata detalhes da consulta
type Metadata struct {
	TotalCount int32 `json:"total_count,omitempty"`
}

//NewMonitor cria um novo host
func NewMonitor(metric Metric) int64 {
	db := Connect()
	defer db.Close()

	//Init transaction
	tx, _ := db.Begin()

	// Insert some data into table.
	sqlStatement, err := db.Prepare("INSERT INTO monitors(hostId, monitorDate, url, statusCod, status, timeResponse) VALUES (?,?,?, ?, ?, ?);")
	res, err := sqlStatement.Exec(&metric.hostId, &metric.monitorDate, &metric.url, &metric.statusCod, &metric.status, &metric.timeResponse)
	fmt.Println(res)

	if err != nil {
		tx.Rollback()
		log.Fatal(err)
		return 0
	}
	id, _ := res.LastInsertId()
	tx.Commit()

	return id
}

// CountMonitor retorna o total de registro
func CountMonitor() (count int) {
	db := Connect()
	defer db.Close()
	rows, _ := db.Query(`SELECT COUNT(*) AS count FROM monitors`)
	for rows.Next() {
		err := rows.Scan(&count)
		if err != nil {
			log.Fatal(err)
		}
	}
	return count
}
