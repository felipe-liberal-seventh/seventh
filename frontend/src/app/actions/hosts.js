import { List } from 'immutable';
import * as SeventhApi from '../../api/SeventhApi';
import * as actionsTypes from './actionsTypes';

export function getHosts() {
    return dispatch => {
        SeventhApi.getHosts()
            .then(hosts => {
                    dispatch({ type: actionsTypes.GET_HOSTS, hosts: new List(hosts.data) })
                    return new List(hosts.data);
                }
            ).catch(error => {
                console.log(error)
            });
    }
}
export function addHost(host) {
     host["protocol"] =  host["ssl"] ? "https://":"http://";
    delete host.ssl
    return dispatch => {
        SeventhApi.postHost(host)
            .then(metric =>{
                dispatch({ type: actionsTypes.ADD_HOST, hosts: new List(metric.data) })
                new List(metric.data)
            }).catch(error => {
                console.log(error)
            });
    };
}
export function deleteHost(host) {
    return dispatch => {
        SeventhApi.deleteHost(host) 
            .then(metric => {
                    dispatch({ type: actionsTypes.DELETE_HOST, hosts: new List(metric.data) })
                    new List(metric.data)
                }
            ).catch(error => {
                console.log(error)
            });
    }
}


