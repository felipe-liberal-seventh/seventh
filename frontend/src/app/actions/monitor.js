import * as SeventhApi from '../../api/SeventhApi';
import * as actionsTypes from './actionsTypes';

export function getMonitor(newHost) {
    console.log("call getMonitor", newHost)
    return dispatch => {
            SeventhApi.getMonitor(newHost)
            .then(metric => {
                console.log(metric)
                    dispatch({ type: actionsTypes.GET_MONITOR, newHost: metric.data })
                    return metric.data;
                }
            ).catch(error => {
                console.log(error)
            });
    }
}
