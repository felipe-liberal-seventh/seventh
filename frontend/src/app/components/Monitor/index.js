import React, { Component } from 'react';
import { getMonitor } from '../../actions/monitor';
import { connect } from 'react-redux';
//import { containerChart} from './styles';
import Highcharts from 'highcharts';
import HighchartsReact from 'highcharts-react-official';
//import HC_more from "highcharts/highcharts-more"; //module
//HC_more(Highcharts); //init module


class Monitor extends Component {
    constructor(props){
        super(props);

        this.allowChartUpdate = true;
        this.state = {};
        
    }
    componentDidMount() {
        console.log('componentDidMount')
        this.props.getMonitor({id:0})
    }
    render() {
        console.log('render')
       
        const { host } = this.props;

          try {        
            console.log(this)

            this.chartData = host.results.map(result => [ new Date(result["monitorDate"]).getTime(), result["timeResponse"] ,result["status"], result["statusCod"], result["url"] ]);
            this.chartName = host.name;
            console.log(new Date(host.results[0]["monitorDate"]));
            console.log(host.results[0]["monitorDate"]);
            }
          catch(err) {
            console.log('aguardando props');
          }
       

       const options = {
            series: [{
                type: 'area',
                name: this.chartName,
                data: this.chartData,
                tooltip:{
                    dateTimeLabelFormats:{
                        hour: "%A, %b %e, %H:%M",
                        day: "%A, %b %e, %H:%M"
                    }
                }
               
            }],
            chart: {
              zoomType: 'x',
              spacingLeft: 30,
              spacingRight: 30,
           
            },
            
            tooltip: {
                
                formatter: function(){
                       return `<span style="font-size:10px"><b>${this.series.name}</b></span> <i><span style="font-size:10px">${new Date(this.point.x).toLocaleDateString('pt-BR')} ${new Date(this.point.x).toLocaleTimeString('pt-BR')}</span></i><br/> 
                                <span style="font-size:10px">${this.point.url}</span><br/> 
                                <span style="font-size:10px">Status: 
                                ${(this.point.statusCod < 200 || this.point.statusCod >= 400) ? '<span style="color:red">' : '<span>'}
                                ${this.point.status}</span><br/>
                                <span style="font-size:12px">Tempo de resposta: <b>${this.point.y}</b>ms</span>`    
                },  
                useHTML: true
            },
            rangeSelector: {
                allButtonsEnabled: true,
                buttons: [{
                    type: 'month',
                    count: 3,
                    text: 'Day',
                    dataGrouping: {
                        forced: true,
                        units: [['day', [1]]]
                    }
                }, {
                    type: 'year',
                    count: 1,
                    text: 'Week',
                    dataGrouping: {
                        forced: true,
                        units: [['week', [1]]]
                    }
                }, {
                    type: 'all',
                    text: 'Month',
                    dataGrouping: {
                        forced: true,
                        units: [['month', [1]]]
                    }
                }],
                buttonTheme: {
                    width: 60
                },
                selected: 2
            },
            _navigator: {
                enabled: false
            },
            subtitle: {
                text: document.ontouchstart === undefined ?
                    'Clique e arraste na Ã¡rea de plotagem para ampliar':'Clique no grafico para amplias'
            },
            time: {useUTC: true},
            xAxis: {
                type: 'datetime'
            },
            yAxis: {
                min: 0,
                labels : { format : '{value:,.0f} ms' },
                title:{ text: 'milisseconds' },          
            },
            credits: {enabled: false},
            title: {
                text: 'Host Monitorados - Tempo de resposta (ms)'
            },
           
            plotOptions: {
                area: {
                    fillColor: {
                        linearGradient: {
                            x1: 0,
                            y1: 0,
                            x2: 0,
                            y2: 1
                        },
                        stops: [
                            [0, Highcharts.getOptions().colors[0]],
                            [1, Highcharts.color(Highcharts.getOptions().colors[0]).setOpacity(0).get('rgba')]
                        ]
                    },
                    marker: { radius: 2 },
                    lineWidth: 1,
                    states: { hover: { lineWidth: 1 } },
                    threshold: null
                },
                series: {
                    keys: ['x', 'y', 'status', 'statusCod','url'], // 4th data position as custom property
                    pointIntervalUnit: 'day'
                }
            },
    }
        //console.log({monitor})
        return (
            <HighchartsReact
            ref={"chartComponent"}
            allowChartUpdate={this.allowChartUpdate}
            highcharts={Highcharts}
            options={ options }
          />

        )
    }
}

const mapStateToProps = function({host}){ 
    return {host}
};
const mapDispatchToProps = dispatch => {
    return {
        getMonitor: (host) => dispatch(getMonitor(host))
    };
}

const MonitorContainer = connect(mapStateToProps, mapDispatchToProps)(Monitor);
export default MonitorContainer;