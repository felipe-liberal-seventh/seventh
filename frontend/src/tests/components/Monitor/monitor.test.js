/**
 * @jest-environment jsdom
 */
import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { shallow, configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Monitor from '../../../app/components/Monitor';
import { createStore, applyMiddleware } from 'redux';
import * as actionsTypes from '../../../app/actions/actionsTypes';
import { Provider } from 'react-redux';
import thunkMiddleware from 'redux-thunk';
import rootReducer from '../../../app/reducers';
const store = createStore(rootReducer, applyMiddleware(thunkMiddleware));
configure({ adapter: new Adapter() });

describe('MonitorComponent', () => {
    const renderer = new ShallowRenderer();

    describe('Snapshots', () => {
      it('Should be a simple MonitorComponent with loadingTemplate', () => {
        const component = renderer.render(
            <Provider store={store}>
                <Monitor />
            </Provider>,
        );
        expect(component).toMatchSnapshot();
      });
    });

    describe('Functions and Branches', () => {
        it('Should be a MonitorComponent with loadingTemplate', () => {
            const component = mount(shallow(
                <Provider store={store}>
                    <Monitor />
                </Provider>
            ).get(0));

            expect(component.find('Monitor img').length).toEqual(1);
        });

        it('Should be a MonitorComponent with monitorTemplate', () => {
            const host = {
                hosts: ['google'],
                value: 'The example text'
            };
            store.dispatch({ type: actionsTypes.GET_MONITOR, monitor });

            const component = mount(shallow(
                <Provider store={store}>
                    <Monitor />
                </Provider>
            ).get(0));

            expect(component.find('.modal-title').text()).toEqual('google');
        });

        it('Should be execute getAnotherJoke function and change joke value', () => {
            let host = {
                hosts: ['google'],
                value: 'The test text'
            };
            store.dispatch({ type: actionsTypes.GET_MONITOR, monitor });

            const component = mount(shallow(
                <Provider store={store}>
                    <Monitor />
                </Provider>
            ).get(0));

            expect(store.getState().joke.value).toEqual('The test text');

            store.dispatch({
                type: actionsTypes.GET_MONITOR,
                host: Object.assign(monitor, { value: 'Another host' })
            });

            const spyGetJoke = spyBuilder(component,'JokeModal', 'getJoke');
            const spyGetAnotherJoke = spyBuilder(component,'JokeModal', 'getAnotherJoke');
            const spyGetFirstCategory = spyBuilder(component,'JokeModal', 'getFirstCategory');
            component.find('#loadAnotherJoke').simulate('click');

            expect(spyGetJoke).toBeCalled();
            expect(spyGetAnotherJoke).toBeCalled();
            expect(spyGetFirstCategory).toBeCalled();
            expect(store.getState().joke.value).toEqual('Another joke');
        });

        it('Should be execute getJokePreviousCategory function with previous categories argument value', () => {
            const categories = ['animal', 'career'];
            const joke = {
                categories: ['career'],
                value: 'The joke text'
            };
            store.dispatch({ type: actionsTypes.GET_JOKE, joke });
            store.dispatch({ type: actionsTypes.GET_CATEGORIES, categories });

            const component = mount(shallow(
                <Provider store={store}>
                    <JokeModal />
                </Provider>
            ).get(0));

            const spyGetJokePreviousCategory = spyBuilder(component,'JokeModal', 'getJokePreviousCategory');
            const spyGetCategoryByIndex = spyBuilder(component,'JokeModal', 'getCategoryByIndex');
            const spyGetJoke = spyBuilder(component,'JokeModal', 'getJoke');
            component.find('#previousJokeCategory').simulate('click');

            expect(spyGetCategoryByIndex).toBeCalled();
            expect(spyGetJokePreviousCategory).toBeCalled();
            expect(spyGetJoke).toHaveBeenCalledWith('animal');
        });

        it('Should be execute getJokeNextCategory function with next categories argument value', () => {
            const categories = ['animal', 'career'];
            const joke = {
                categories: ['career'],
                value: 'The joke text'
            };
            store.dispatch({ type: actionsTypes.GET_JOKE, joke });
            store.dispatch({ type: actionsTypes.GET_CATEGORIES, categories });

            const component = mount(shallow(
                <Provider store={store}>
                    <JokeModal />
                </Provider>
            ).get(0));

            const spyGetJokeNextCategory = spyBuilder(component,'JokeModal', 'getJokeNextCategory');
            const spyGetCategoryByIndex = spyBuilder(component,'JokeModal', 'getCategoryByIndex');
            const spyGetJoke = spyBuilder(component,'JokeModal', 'getJoke');
            component.find('#nextJokeCategory').simulate('click');

            expect(spyGetCategoryByIndex).toBeCalled();
            expect(spyGetJokeNextCategory).toBeCalled();
            expect(spyGetJoke).toHaveBeenCalledWith('animal');
        });

        it('Should be execute getJokePreviousCategory function and turn to the last category value', () => {
            const categories = ['animal', 'career'];
            const joke = {
                categories: ['animal'],
                value: 'The joke text'
            };
            store.dispatch({ type: actionsTypes.GET_JOKE, joke });
            store.dispatch({ type: actionsTypes.GET_CATEGORIES, categories });

            const component = mount(shallow(
                <Provider store={store}>
                    <JokeModal />
                </Provider>
            ).get(0));

            const spyGetJoke = spyBuilder(component,'JokeModal', 'getJoke');
            component.find('#previousJokeCategory').simulate('click');

            expect(spyGetJoke).toHaveBeenCalledWith('career');
        });
    });
});

function spyBuilder(component, wrapperName, method) {
    return jest.spyOn(component.find(wrapperName).instance(), method);
}