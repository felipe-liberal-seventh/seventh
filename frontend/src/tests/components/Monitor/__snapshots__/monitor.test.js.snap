// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonitorComponent Snapshots Should be a simple MonitorComponent with loadingTemplate 1`] = `
<Context.Provider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <UNDEFINED />
</Context.Provider>
`;
